### PR TITLE
feat: support openai audio api

### PR DESCRIPTION
**Pull request by aiwantaozi**
https://github.com/gpustack/gpustack/issues/428

### CODE REVIEW

File: gpustack/routes/models.py [hunk 1]
```diff
@@ -114,3 +118,20 @@ async def delete_model(session: SessionDep, id: int):
...
+@router.get("/{id}/voices", response_model=VoicesResponse)
+async def voices(session: SessionDep, request: Request, id: int):
+    model = await Model.one_by_id(session, id)
+    if not model:
+        raise NotFoundException(message="Model not found")
+
+    if model.tts is None or not model.tts:
+        raise BadRequestException(message="Model does not support voices API")
+
+    return await proxy_request_by_model(request, session, "voices")
```
Comment (gitlawr) [inline gpustack/routes/models.py:165]: Any documentation for this backend API?

File: gpustack/routes/openai.py [hunk 2]
```diff
@@ -95,28 +106,76 @@ async def proxy_request_by_model(  # noqa: C901
...
             is_openai_exception=True,
         )
 
-    model_name = body.get("model")
-    if not model_name:
+    model_name = None
+    model_id = None
+    if request.method == "GET":
+        model_id = request.path_params.get("id")
+    elif content_type.startswith("application/json"):
```
Comment (gitlawr) [inline gpustack/routes/openai.py]: Now it requires the content-type in request headers but it's optional before

File: gpustack/routes/openai.py [hunk 3]
```diff
@@ -177,7 +237,21 @@ async def stream_generator():
                     method=request.method,
                     url=url,
                     headers=headers,
-                    json=body,
+                    json=(
+                        body_json
+                        if content_type.startswith("application/json")
+                        else None
+                    ),
+                    data=(
```
Comment (gitlawr) [inline gpustack/routes/openai.py]: Should we have a different proxy function to handle the form data case?

File: gpustack/scheduler/scheduler.py [hunk 4]
```diff
@@ -338,9 +334,9 @@ async def _schedule_one(self, instance: ModelInstance):
         state_message = ""
 
         async with AsyncSession(self._engine) as session:
-            workers = await Worker.all(session)
+            workers = await Worker.all_by_field(session, "state", WorkerStateEnum.READY)
```
Comment (gitlawr) [inline gpustack/scheduler/scheduler.py]: Keep the origin, so that StatusFilter can work and provide useful info

File: gpustack/schemas/models.py [hunk 5]
```diff
@@ -92,6 +94,8 @@ class ModelBase(SQLModel, ModelSource):
     gpu_selector: Optional[GPUSelector] = Field(
         sa_column=Column(pydantic_column_type(GPUSelector)), default=None
     )
+    stt: bool = False
+    tts: bool = False
```
Comment (gitlawr) [inline gpustack/schemas/models.py]: Use full form

File: gpustack/schemas/models.py [hunk 6]
```diff
@@ -266,11 +275,23 @@ def is_gguf_model(model: Model):
     )
 
 
+def is_stt_tts_model(model: Model):
```
Comment (gitlawr) [inline gpustack/schemas/models.py]: is `audio_model` better?

File: pyproject.toml [hunk 7]
```diff
@@ -48,7 +48,7 @@ wmi = { version="^1.5.1", markers = "platform_system == 'Windows'" }
 pywin32 = { version="^308", markers = "platform_system == 'Windows'" }
 packaging = "^24.1"
 psycopg2-binary = "^2.9.10"
-
+pipx = "^1.7.1"
```
Comment (gitlawr) [inline pyproject.toml]: Why this dependency

File: gpustack/migrations/versions/2024_11_04_1347-004c73a5c09e_add_local_path.py [hunk 8]
```diff
@@ -34,9 +34,18 @@ def upgrade() -> None:
...
         # The default is False, but is not valid in SQLite which uses interger for boolean. Use a generic representation.
         batch_op.add_column(sa.Column('image_only', sa.Boolean(), nullable=False, server_default="0"))
+        batch_op.add_column(sa.Column('speech_to_text', sa.Boolean(),
+                            nullable=False, server_default="0"))
+        batch_op.add_column(sa.Column('text_to_speech', sa.Boolean(),
+                            nullable=False, server_default="0"))
 
 
     with op.batch_alter_table('model_usages', naming_convention=naming_convention) as batch_op:
+        batch_op.alter_column('operation',
```
Comment (gitlawr) [inline gpustack/migrations/versions/2024_11_04_1347-004c73a5c09e_add_local_path.py:44]: Can you check if the upgrade works in pgsql

File: gpustack/routes/models.py (hunk 1 above)
Comment (aiwantaozi) [reply gpustack/routes/models.py:165]: Get `/v1/voices`
```json
{
    "voices": [
        "中文女",
        "中文男",
        "日语男",
        "粤语女",
        "英文女",
        "英文男",
        "韩语女"
    ]
}
```

**Pull request merged**